### PR TITLE
[FLINK-27790][Table SQL / API] Port ADD JAR /SHOW JARS syntax implementation from SqlClient to TableEnvironment side

### PR DESCRIPTION
**Pull request by Myracle**
## What is the purpose of the change

*Port ADD JAR /SHOW JARS syntax implementation from SqlClient to TableEnvironment side for more common use.*

## Brief change log

  - *Add methods addJar and listJars to the TableEnvironment.*
  - *Call the methods from SessionContext in SQL client.*

## Verifying this change

This change added tests and can be verified as follows:

  - *Added ITCase AddRemoteJarITCase.*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)

### CODE REVIEW

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/TableEnvironmentImpl.java [hunk 1]
```diff
@@ -461,6 +463,23 @@ public boolean dropTemporaryFunction(String path) {
         return functionCatalog.dropTemporaryCatalogFunction(unresolvedIdentifier, true);
     }
 
+    @Override
+    public void addJar(String jarPath) {
+        try {
+            resourceManager.registerJarResources(
```
Comment (beyond1920) [inline flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/TableEnvironmentImpl.java:469]: Do we need save jar into catalog?
Or added Jar always work in session scope?

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/TableEnvironmentImpl.java [hunk 2]
```diff
@@ -461,6 +463,23 @@ public boolean dropTemporaryFunction(String path) {
...
+                    Collections.singletonList(new ResourceUri(ResourceType.JAR, jarPath)));
+        } catch (IOException e) {
+            throw new TableException(String.format("Could not register the specified jar [%s].", jarPath), e);
+        }
+    }
+
+    @Override
+    public List<String> listJars() {
```
Comment (beyond1920) [inline flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/TableEnvironmentImpl.java:477]: AddJar and listJars could work for DataStream user or TableAPI user.
How to use this function for SQL user?
We also need to support `AddJarOperation` in `executeInternal` method, WDYT?

**Review by beyond1920 — COMMENTED** [review]
Thanks for contribution.
I left some comments.

**Review by lsyldliu — COMMENTED** [review]
@Myracle Thanks for contribution, I left some comments.

File: flink-end-to-end-tests/flink-end-to-end-tests-sql/src/test/java/org/apache/flink/table/sql/codegen/AddRemoteJarITCase.java [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+import java.util.List;
+import java.util.Map;
+import java.util.UUID;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.junit.Assert.assertTrue;
+
+/** ITCase for adding remote jar. */
+@RunWith(Parameterized.class)
+public class AddRemoteJarITCase extends TestLogger {
```
Comment (lsyldliu) [inline flink-end-to-end-tests/flink-end-to-end-tests-sql/src/test/java/org/apache/flink/table/sql/codegen/AddRemoteJarITCase.java:70]: Please use junit5 for new added test, refer to https://flink.apache.org/contributing/code-style-and-quality-common.html#testing

File: flink-end-to-end-tests/flink-end-to-end-tests-sql/src/test/java/org/apache/flink/table/sql/codegen/AddRemoteJarITCase.java [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+
+            List<String> sqlLines = initializeSqlLines(varsMap);
+
+            executeSqlStatements(clusterController, sqlLines);
+
+            checkJsonResultFile();
+        }
+    }
+
+    private List<String> initializeSqlLines(Map<String, String> vars) throws IOException {
```
Comment (lsyldliu) [inline flink-end-to-end-tests/flink-end-to-end-tests-sql/src/test/java/org/apache/flink/table/sql/codegen/AddRemoteJarITCase.java:185]: Here many duplicated method with `PlannerScalaFreeITCase`, so I think we can extract an base class to reuse the common method.

File: flink-end-to-end-tests/flink-end-to-end-tests-sql/src/test/java/org/apache/flink/table/sql/codegen/AddRemoteJarITCase.java (hunk 3 above)
Comment (lsyldliu) [inline flink-end-to-end-tests/flink-end-to-end-tests-sql/src/test/java/org/apache/flink/table/sql/codegen/AddRemoteJarITCase.java:70]: Change this class name to `UsingRemoteJarITCase`? We also add `create function ... using jar ...` related tests in the future.

File: flink-end-to-end-tests/flink-end-to-end-tests-sql/src/test/java/org/apache/flink/table/sql/codegen/AddRemoteJarITCase.java [hunk 5]
```diff
@@ -0,0 +1,280 @@
...
+    private static final Logger LOG = LoggerFactory.getLogger(AddRemoteJarITCase.class);
+
+    private static final String ADD_REMOTE_JAR_E2E_SQL = "add_remote_jar_e2e.sql";
+
+    private static final String EXECUTE_SQL_RESULT =
+            "{\"before\":null,\"after\":{\"id\":1,\"content\":\"hello world\"},\"op\":\"c\"}";
+
+    public static final String GENERATED_LOWER_UDF_CLASS = "LowerUDF";
+
+    public static final String GENERATED_LOWER_UDF_CODE =
```
Comment (lsyldliu) [inline flink-end-to-end-tests/flink-end-to-end-tests-sql/src/test/java/org/apache/flink/table/sql/codegen/AddRemoteJarITCase.java:80]: Please implement this udf in **flink-sql-client-test** module, it will be packaged into SqlToolbox.jar, then we can use it directly.

File: flink-end-to-end-tests/flink-end-to-end-tests-sql/src/test/java/org/apache/flink/table/sql/codegen/AddRemoteJarITCase.java [hunk 6]
```diff
@@ -0,0 +1,280 @@
...
+            hdfsCluster.shutdown();
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    @Test
+    public void testAddRemoteJar() throws Exception {
+        String remoteJarPath =
+                "hdfs://"
```
Comment (lsyldliu) [inline flink-end-to-end-tests/flink-end-to-end-tests-sql/src/test/java/org/apache/flink/table/sql/codegen/AddRemoteJarITCase.java:164]: Using String.format()

File: flink-end-to-end-tests/flink-end-to-end-tests-sql/src/test/java/org/apache/flink/table/sql/codegen/AddRemoteJarITCase.java [hunk 7]
```diff
@@ -0,0 +1,280 @@
...
+    }
+
+    @Rule
+    public final FlinkResource flink =
+            new LocalStandaloneFlinkResourceFactory().create(FlinkResourceSetup.builder().build());
+
+    private final String executionMode;
+    private Path result;
+
+    private MiniDFSCluster hdfsCluster;
```
[the remark this answers is not in the document]
Comment (Myracle) [reply flink-end-to-end-tests/flink-end-to-end-tests-sql/src/test/java/org/apache/flink/table/sql/codegen/AddRemoteJarITCase.java:105]: I agree with you that a more common way is benefit. But the related tests with  HbaseResource and KafkaResource  are all ignored for that the tests maybe get stuck. If we download hdfs and setup a local hdfs cluster by command, I am afraid the process maybe get stuck someday. Since many ITCase use MiniDFSCluster, current implementation is a stable test. What do you think?

File: flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/api/TableEnvironmentTest.scala [hunk 8]
```diff
@@ -155,6 +164,23 @@ class TableEnvironmentTest {
...
+        "test-classloader-udf.jar",
+        GENERATED_LOWER_UDF_CLASS,
+        String.format(GENERATED_LOWER_UDF_CODE, GENERATED_LOWER_UDF_CLASS)
+      )
+      .getPath
+
+    tableEnv.executeSql("add JAR '" + jarPath + "'")
+    val tableResult = tableEnv.executeSql("SHOW JARS")
+
+    checkData(util.Arrays.asList(Row.of(jarPath)).iterator(), tableResult.collect())
```
[the remark this answers is not in the document]
Comment (Myracle) [reply flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/api/TableEnvironmentTest.scala:181]: It is ok to assert the ResultKind. Since there is already a method checkData, why not just use it?

File: flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/gateway/context/SessionContext.java [hunk 9]
```diff
@@ -248,6 +248,7 @@ public static SessionContext create(DefaultContext defaultContext, String sessio
                 executionContext);
     }
 
+    // TODO: Only related to removeJar in test. Remove it once it has no use.
```
[the remark this answers is not in the document]
Comment (Myracle) [reply flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/gateway/context/SessionContext.java:251]: It is a choice. But the removeJar needs add jar firstly. If the addJar method is removed, I has to construct another addJar in sessionContestTest. What do you think about this way?

File: flink-end-to-end-tests/flink-end-to-end-tests-sql/src/test/java/org/apache/flink/table/sql/codegen/AddRemoteJarITCase.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (lsyldliu) [reply flink-end-to-end-tests/flink-end-to-end-tests-sql/src/test/java/org/apache/flink/table/sql/codegen/AddRemoteJarITCase.java]: Ok, I think your concern make sense to me.

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/TableEnvironmentImpl.java (hunk 2 above)
Comment (Myracle) [reply flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/TableEnvironmentImpl.java]: As @lsyldliu, the public api will be discussed later.

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/TableEnvironmentImpl.java (hunk 1 above)
Comment (Myracle) [reply flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/TableEnvironmentImpl.java]: It is not related to catalog.

File: flink-table/flink-sql-client/src/test/java/org/apache/flink/table/client/gateway/context/SessionContextTest.java [hunk 10]
```diff
@@ -156,13 +156,7 @@ public void testAddJarWithRelativePath() throws IOException {
 
     @Test
     public void testAddIllegalJar() {
-        validateAddJarWithException("/path/to/illegal.jar", "JAR file does not exist");
-    }
-
-    @Test
-    public void testAddRemoteJar() {
-        validateAddJarWithException(
```
[the remark this answers is not in the document]
Comment (Myracle) [reply flink-table/flink-sql-client/src/test/java/org/apache/flink/table/client/gateway/context/SessionContextTest.java:164]: The e2e test UsingRemoteJarITCase will replace it.

File: flink-end-to-end-tests/flink-end-to-end-tests-common/src/main/java/org/apache/flink/tests/util/flink/FlinkDistribution.java [hunk 11]
```diff
@@ -220,6 +220,7 @@ public void submitSQLJob(SQLJobSubmission job, Duration timeout) throws IOExcept
         AutoClosableProcess.create(commands.toArray(new String[0]))
                 .setStdInputs(job.getSqlLines().toArray(new String[0]))
                 .setStdoutProcessor(LOG::info) // logging the SQL statements and error message
+                .setEnv(job.getEnvProcessor())
```
[the remark this answers is not in the document]
Comment (Myracle) [reply flink-end-to-end-tests/flink-end-to-end-tests-common/src/main/java/org/apache/flink/tests/util/flink/FlinkDistribution.java:223]: It is needed in UsingRemoteJarITCase.

File: flink-end-to-end-tests/flink-end-to-end-tests-sql/src/test/java/org/apache/flink/table/sql/codegen/AddRemoteJarITCase.java [hunk 12]
```diff
@@ -0,0 +1,280 @@
...
+                "HDFS cluster cannot be started on Windows without extensions.",
+                !OperatingSystem.isWindows());
+    }
+
+    @Before
+    public void createHDFS() {
+        try {
+            Configuration hdConf = new Configuration();
+
+            File baseDir = new File("./target/hdfs/hdfsTest").getAbsoluteFile();
```
[the remark this answers is not in the document]
Comment (Myracle) [reply flink-end-to-end-tests/flink-end-to-end-tests-sql/src/test/java/org/apache/flink/table/sql/codegen/AddRemoteJarITCase.java]: The subDir of current test is code path. I prefer to use target.

File: flink-end-to-end-tests/flink-end-to-end-tests-sql/src/test/java/org/apache/flink/table/sql/codegen/AddRemoteJarITCase.java [hunk 13]
```diff
@@ -0,0 +1,280 @@
...
+    private org.apache.hadoop.fs.Path hdPath;
+    private org.apache.hadoop.fs.FileSystem hdfs;
+
+    public AddRemoteJarITCase(String executionMode) {
+        this.executionMode = executionMode;
+    }
+
+    @BeforeClass
+    public static void verifyOS() {
+        Assume.assumeTrue(
```
[the remark this answers is not in the document]
Comment (Myracle) [reply flink-end-to-end-tests/flink-end-to-end-tests-sql/src/test/java/org/apache/flink/table/sql/codegen/AddRemoteJarITCase.java]: Please refer to https://issues.apache.org/jira/browse/FLINK-6558

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/runtime/batch/sql/FunctionITCase.java [hunk 14]
```diff
@@ -142,6 +147,23 @@ public void testUserDefinedTemporaryCatalogFunctionByUsingJar() throws Exception
...
+        assertThat(
+                        CollectionUtil.iteratorToList(tableResult.collect())
+                                .equals(
+                                        Collections.singletonList(
+                                                Row.of(new Path(jarPath).getPath()))))
+                .isTrue();
+
+        testUserDefinedFunctionByUsingJar(
+                String.format("create function lowerUdf as '%s' LANGUAGE JAVA", udfClassName),
+                null);
```
Comment (lsyldliu) [inline flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/runtime/batch/sql/FunctionITCase.java:164]: why not delete the function?

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/runtime/stream/sql/FunctionITCase.java [hunk 15]
```diff
@@ -1225,6 +1228,23 @@ public void testWithBoolNotNullTypeHint() {
...
+        assertThat(
+                        CollectionUtil.iteratorToList(tableResult.collect())
+                                .equals(
+                                        Collections.singletonList(
+                                                Row.of(new Path(jarPath).getPath()))))
+                .isTrue();
+
+        testUserDefinedFunctionByUsingJar(
+                String.format("create function lowerUdf as '%s' LANGUAGE JAVA", udfClassName),
+                null);
```
Comment (lsyldliu) [inline flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/runtime/stream/sql/FunctionITCase.java:1245]: ditto

File: flink-end-to-end-tests/flink-end-to-end-tests-sql/src/test/resources/log4j2-test.properties [hunk 16]
```diff
@@ -18,7 +18,7 @@
 
 # Set root logger level to OFF to not flood build logs
 # set manually to INFO for debugging purposes
-rootLogger.level = OFF
+rootLogger.level = INFO
```
Comment (lsyldliu) [inline flink-end-to-end-tests/flink-end-to-end-tests-sql/src/test/resources/log4j2-test.properties:21]: Please revert this change.

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/runtime/batch/sql/FunctionITCase.java (hunk 14 above)
Comment (Myracle) [reply flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/runtime/batch/sql/FunctionITCase.java:164]: Using this function to verify the execute result.

File: flink-test-utils-parent/flink-test-utils/src/main/java/org/apache/flink/test/util/SQLJobSubmission.java [hunk 17]
```diff
@@ -29,10 +31,13 @@ public class SQLJobSubmission {
 
     private final List<String> sqlLines;
     private final List<String> jars;
+    private final Consumer<Map<String, String>> envProcessor;
```
[the remark this answers is not in the document]
Comment (Myracle) [reply flink-test-utils-parent/flink-test-utils/src/main/java/org/apache/flink/test/util/SQLJobSubmission.java:35]: This is needed for UsingRemoteJarITCaseBase

File: flink-end-to-end-tests/flink-end-to-end-tests-sql/src/test/java/org/apache/flink/table/sql/codegen/UsingRemoteJarITCaseBase.java [hunk 18]
```diff
@@ -0,0 +1,162 @@
...
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.nio.file.Path;
+import java.time.Duration;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+
+/** ITCase for adding remote jar. */
+public class UsingRemoteJarITCaseBase extends SqlITCaseBase {
```
Comment (lsyldliu) [inline flink-end-to-end-tests/flink-end-to-end-tests-sql/src/test/java/org/apache/flink/table/sql/codegen/UsingRemoteJarITCaseBase.java:47]:
```suggestion
public class UsingRemoteJarITCase extends SqlITCaseBase {
```

File: flink-end-to-end-tests/flink-end-to-end-tests-sql/src/test/java/org/apache/flink/table/sql/codegen/UsingRemoteJarITCaseBase.java [hunk 19]
```diff
@@ -0,0 +1,162 @@
...
+                    "{\"before\":null,\"after\":{\"id\":1,\"content\":\"Hello Flink\"},\"op\":\"c\"}");
+
+    private static final Path HADOOP_CLASSPATH = TestUtils.getResource(".*hadoop.classpath");
+
+    private MiniDFSCluster hdfsCluster;
+    private org.apache.hadoop.fs.Path hdPath;
+    private org.apache.hadoop.fs.FileSystem hdfs;
+
+    public UsingRemoteJarITCaseBase(String executionMode) {
+        super(executionMode, new org.apache.flink.configuration.Configuration());
```
Comment (lsyldliu) [inline flink-end-to-end-tests/flink-end-to-end-tests-sql/src/test/java/org/apache/flink/table/sql/codegen/UsingRemoteJarITCaseBase.java:61]: Why here need to new a default `Configuration ` instead of reuse the configuration in parent class?

File: flink-end-to-end-tests/flink-end-to-end-tests-sql/src/test/java/org/apache/flink/table/sql/codegen/SqlITCaseBase.java [hunk 20]
```diff
@@ -0,0 +1,178 @@
...
+            throws Exception {
+        try (ClusterController clusterController = flink.startCluster(1)) {
+            List<String> sqlLines = initializeSqlLines(sqlPath, varsMap);
+
+            executeSqlStatements(clusterController, sqlLines);
+
+            // Wait until all the results flushed to the json file.
+            LOG.info("Verify the json result.");
+            checkJsonResultFile(resultSize, resultItems);
+            LOG.info("The codegen SQL client test run successfully.");
```
Comment (lsyldliu) [inline flink-end-to-end-tests/flink-end-to-end-tests-sql/src/test/java/org/apache/flink/table/sql/codegen/SqlITCaseBase.java:104]:
```suggestion
            LOG.info("The SQL client test run successfully.");
```

File: flink-end-to-end-tests/flink-end-to-end-tests-sql/src/test/java/org/apache/flink/table/sql/codegen/PlannerScalaFreeITCaseBase.java [hunk 21]
```diff
@@ -0,0 +1,70 @@
...
+import java.util.Arrays;
+import java.util.List;
+
+/**
+ * End to End tests for table planner scala-free since 1.15. Due to scala-free of table planner
+ * introduced, the class in table planner is not visible in distribution runtime, if we use these
+ * class in execution time, ClassNotFound exception will be thrown. ITCase in table planner can not
+ * cover it, so we should add E2E test for these case.
+ */
+public class PlannerScalaFreeITCaseBase extends SqlITCaseBase {
```
Comment (lsyldliu) [inline flink-end-to-end-tests/flink-end-to-end-tests-sql/src/test/java/org/apache/flink/table/sql/codegen/PlannerScalaFreeITCaseBase.java:37]:
```suggestion
public class PlannerScalaFreeITCase extends SqlITCaseBase {
```

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/runtime/batch/sql/FunctionITCase.java (hunk 14 above)
Comment (lsyldliu) [reply flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/runtime/batch/sql/FunctionITCase.java:164]: Please delete it, we has verified the result before drop it.

File: flink-end-to-end-tests/flink-end-to-end-tests-sql/src/test/java/org/apache/flink/table/sql/codegen/AddRemoteJarITCase.java (hunk 3 above)
Comment (Myracle) [reply flink-end-to-end-tests/flink-end-to-end-tests-sql/src/test/java/org/apache/flink/table/sql/codegen/AddRemoteJarITCase.java]: Some features written in junit4 can not be used in junit5. So we decide not to use junit5 for this test.

**Comment by Myracle** [comment]
@flinkbot run azure

**Review by lsyldliu — APPROVED** [review]
LGTM

**Review by wuchong — COMMENTED** [review]
The `function.q` of SQL Client is failed. Please take a look.

**Review by wuchong — APPROVED** [review]
I have appended commits to fix the concerns. Will merge it once CI is passed.

**Comment by wuchong** [comment]
The failed case is tracked by FLINK-28844.
Merging...
